File: lib/dreaming/dream-diary.js
import { createHash } from "node:crypto";
import { lstatSync, mkdtempSync, readFileSync, renameSync, rmSync, writeFileSync } from "node:fs";
import { basename, dirname, join, resolve } from "node:path";

import { loadOpenClawPluginSdkRuntime } from "../setup/feature-cron-plugin-runtime.js";

/**
 * Dream diary bridge.
 *
 * OpenClaw's Control UI (Settings → Memory → Dreams → Diary) shows one file:
 * `DREAMS.md` in the agent's workspace, read as a whole through the host's
 * `doctor.memory.dreamDiary` method. The host does not ask the memory plugin
 * for it. So the way for a PLUR1BUS dream to appear on that page is to write
 * it into that file, in the block and the entry shape the host itself uses:
 *
 *   # Dream Diary
 *   <!-- openclaw:dreaming:diary:start -->
 *   ---
 *   *September 4, 2026 at 3:06 AM GMT+2*
 *   <narrative>
 *   <!-- openclaw:dreaming:diary:end -->
 *
 * The host's own maintenance (dedupe, backfill strip, artifact repair) works
 * on fingerprints and on blocks it marked itself, so a foreign entry in this
 * shape survives it. That is read from the host source, not a contract.
 */
export const DREAM_DIARY_FILE_NAMES = Object.freeze(["DREAMS.md", "dreams.md"]);
export const DIARY_HEADING = "# Dream Diary";
export const DIARY_START_MARKER = "<!-- openclaw:dreaming:diary:start -->";
export const DIARY_END_MARKER = "<!-- openclaw:dreaming:diary:end -->";
export const DIARY_MAX_NARRATIVE_CHARS = 8_000;
const DIARY_MAX_FILE_BYTES = 8 * 1024 * 1024;
const MODE_LABELS = Object.freeze({ light: "light dream", rem: "REM dream" });

/** Same shape as the host's own diary timestamp, e.g. "September 4, 2026 at 3:06 AM GMT+2". */
export function formatDiaryDate(epochMs, timezone) {
  const options = {
    year: "numeric",
    month: "long",
    day: "numeric",
    hour: "numeric",
    minute: "2-digit",
    hour12: true,
    timeZoneName: "short",
  };
  if (typeof timezone === "string" && timezone.trim()) options.timeZone = timezone.trim();
  try {
    return new Intl.DateTimeFormat("en-US", options).format(new Date(epochMs));
  } catch {
    delete options.timeZone;
    return new Intl.DateTimeFormat("en-US", options).format(new Date(epochMs));
  }
}

function normalizeNarrative(narrative) {
  if (typeof narrative !== "string") return "";
  const text = narrative.replace(/\r\n?/g, "\n").trim();
  if (!text) return "";
  return text.length > DIARY_MAX_NARRATIVE_CHARS ? `${text.slice(0, DIARY_MAX_NARRATIVE_CHARS).trimEnd()}…` : text;
}

/** Stable identity of one narrative, used to keep a re-run from writing the same dream twice. */
export function narrativeFingerprint(narrative) {
  return createHash("sha256").update(normalizeNarrative(narrative).replace(/\s+/g, " ").toLowerCase()).digest("hex").slice(0, 16);
}

/** One diary entry in the host's shape, with a small source line so the reader can tell who dreamt. */
export function buildDiaryEntry({ narrative, mode = "rem", dateText }) {
  const text = normalizeNarrative(narrative);
  if (!text) throw new Error("dream diary entry needs a narrative");
  const label = MODE_LABELS[mode] || `${String(mode)} dream`;
  return `\n---\n\n*${dateText}*\n\n${text}\n\n<sub>PLUR1BUS · ${label} · ${narrativeFingerprint(text)}</sub>\n`;
}

/**
 * Insert an entry into the diary block of an existing file body.
 * @returns {{content: string, changed: boolean, reason?: string}}
 */
export function insertDiaryEntry(existing, entry, { narrative } = {}) {
  const original = typeof existing === "string" ? existing.replace(/\r\n?/g, "\n") : "";
  const fingerprint = narrativeFingerprint(narrative ?? entry);
  if (original.includes(fingerprint)) return { content: original, changed: false, reason: "already_present" };
  const startIdx = original.indexOf(DIARY_START_MARKER);
  const endIdx = original.lastIndexOf(DIARY_END_MARKER);
  if (startIdx >= 0 && endIdx > startIdx) {
    return { content: `${original.slice(0, endIdx)}${entry}${original.slice(endIdx)}`, changed: true };
  }
  // No diary block yet: put one at the top, the way the host does, and keep
  // whatever else the file already holds below it.
  const section = `${DIARY_HEADING}\n\n${DIARY_START_MARKER}${entry}\n${DIARY_END_MARKER}\n`;
  const rest = original.trim().length === 0 ? "" : `\n${original.replace(/^\n+/, "")}`;
  return { content: `${section}${rest}`, changed: true };
}

function resolveDiaryPath(workspaceDir) {
  const root = resolve(workspaceDir);
  for (const name of DREAM_DIARY_FILE_NAMES) {
    const candidate = join(root, name);
    let stat;
    try {
      stat = lstatSync(candidate);
    } catch (error) {
      if (error?.code === "ENOENT") continue;
      throw error;
    }
    if (stat.isSymbolicLink()) throw new Error("dream diary path is a symbolic link");
    if (!stat.isFile()) throw new Error("dream diary path is not a regular file");
    if (stat.size > DIARY_MAX_FILE_BYTES) throw new Error("dream diary file is too large to update");
    return { path: candidate, exists: true };
  }
  return { path: join(root, DREAM_DIARY_FILE_NAMES[0]), exists: false };
}

function writeAtomically(path, content) {
  const dir = dirname(path);
  const tmpDir = mkdtempSync(join(dir, `.${basename(path)}.tmp-`));
  const tmpFile = join(tmpDir, basename(path));
  try {
    writeFileSync(tmpFile, content, "utf8");
    renameSync(tmpFile, path);
  } finally {
    rmSync(tmpDir, { recursive: true, force: true });
  }
}

/**
 * Append one dream to the agent's diary file. Fail-open by contract: a dream
 * that cannot be written to the diary is still a dream, so this never throws
 * and reports what it did instead.
 * @returns {{written: boolean, path: string|null, reason?: string}}
 */
export function appendDreamDiaryEntry({
  workspaceDir,
  narrative,
  mode = "rem",
  timezone = null,
  now = Date.now,
  logger = null,
} = {}) {
  if (typeof workspaceDir !== "string" || !workspaceDir.trim()) return { written: false, path: null, reason: "no_workspace" };
  const text = normalizeNarrative(narrative);
  if (!text) return { written: false, path: null, reason: "empty_narrative" };
  try {
    const { path, exists } = resolveDiaryPath(workspaceDir);
    const existing = exists ? readFileSync(path, "utf8") : "";
    const entry = buildDiaryEntry({ narrative: text, mode, dateText: formatDiaryDate(now(), timezone) });
    const next = insertDiaryEntry(existing, entry, { narrative: text });
    if (!next.changed) return { written: false, path, reason: next.reason };
    writeAtomically(path, next.content);
    logger?.info?.(`dream-diary[${mode}]: entry appended to ${basename(path)}`);
    return { written: true, path };
  } catch (error) {
    logger?.warn?.(`dream-diary[${mode}]: diary not updated (fail-open): ${error?.message || error}`);
    return { written: false, path: null, reason: "write_failed" };
  }
}

/**
 * Tell the host a dream completed, through its public event log. The Dreams
 * page does not depend on it, but the host's own tooling reads that log, and
 * the shape is the one memory-core itself writes. Fail-open like the diary.
 */
export async function emitDreamCompletedEvent({
  workspaceDir,
  mode = "rem",
  narrative,
  now = Date.now,
  // A bare import of "openclaw/..." does not resolve from the extension
  // directory; the SDK loader finds the active host package first.
  importHostEvents = () => loadOpenClawPluginSdkRuntime("memory-host-events"),
  logger = null,
} = {}) {
  if (typeof workspaceDir !== "string" || !workspaceDir.trim()) return false;
  const phase = mode === "light" ? "light" : "rem";
  const lineCount = normalizeNarrative(narrative).split("\n").filter((line) => line.trim()).length;
  try {
    const host = await importHostEvents();
    if (typeof host?.appendMemoryHostEvent !== "function") return false;
    await host.appendMemoryHostEvent(workspaceDir, {
      type: "memory.dream.completed",
      timestamp: new Date(now()).toISOString(),
      phase,
      outcome: "completed",
      lineCount,
      storageMode: "inline",
    });
    return true;
  } catch (error) {
    logger?.warn?.(`dream-diary[${mode}]: host event not written (fail-open): ${error?.message || error}`);
    return false;
  }
}

File: plur1bus-hermes/src/plur1bus_hermes/dream_diary.py
"""Safe private-agent DREAMS.md managed-block writer."""

from __future__ import annotations

import hashlib
import os
import re
import tempfile
import threading
from datetime import datetime, timezone
from pathlib import Path
from typing import Any
from zoneinfo import ZoneInfo

from .namespaces import binding_from_scope
from .validation import safe_agent_id


DIARY_FILE = "DREAMS.md"
START_MARKER = "<!-- openclaw:dreaming:diary:start -->"
END_MARKER = "<!-- openclaw:dreaming:diary:end -->"
MAX_NARRATIVE_CHARS = 8_000
MAX_FILE_BYTES = 8 * 1024 * 1024
_DIARY_LOCK = threading.RLock()


def _normalize_narrative(value: Any) -> str:
    if not isinstance(value, str):
        return ""
    text = value.replace("\r\n", "\n").replace("\r", "\n").strip()
    if not text:
        return ""
    # A narrative is untrusted text.  It must not be able to close or create
    # the host-managed block that contains it.
    return text[:MAX_NARRATIVE_CHARS].rstrip().replace(
        "<!-- openclaw:dreaming:diary:", "<!-- plur1bus:untrusted-diary-marker:"
    )


def narrative_fingerprint(narrative: Any) -> str:
    """Return a stable short fingerprint used for idempotent diary appends."""
    normalized = re.sub(r"\s+", " ", _normalize_narrative(narrative)).lower()
    return hashlib.sha256(normalized.encode("utf-8")).hexdigest()[:16]


def _date_text(now: datetime | None, timezone_name: str | None) -> str:
    moment = now or datetime.now(timezone.utc)
    if moment.tzinfo is None:
        moment = moment.replace(tzinfo=timezone.utc)
    if timezone_name:
        try:
            moment = moment.astimezone(ZoneInfo(timezone_name))
        except Exception:
            moment = moment.astimezone(timezone.utc)
    return moment.strftime("%B %-d, %Y at %-I:%M %p %Z")


def _workspace_path(workspace_dir: str | Path, workspace_root: str | Path | None) -> Path:
    candidate = Path(workspace_dir).expanduser()
    if not candidate.is_absolute():
        raise ValueError("workspace directory is invalid")
    if workspace_root is None:
        if candidate.is_symlink() or not candidate.is_dir():
            raise ValueError("workspace directory is invalid")
        return candidate.resolve()
    root = Path(workspace_root).expanduser()
    if not root.is_absolute() or root.is_symlink() or not root.is_dir():
        raise ValueError("workspace root is invalid")
    try:
        parts = candidate.relative_to(root).parts
    except ValueError as error:
        raise ValueError("workspace directory escapes its root") from error
    current = root
    for part in parts:
        current = current / part
        if current.is_symlink():
            raise ValueError("workspace directory contains a symbolic link")
        if not current.exists():
            current.mkdir()
        if not current.is_dir():
            raise ValueError("workspace path is not a directory")
    return candidate.resolve()


def _diary_path(workspace: Path) -> tuple[Path, bool]:
    path = workspace / DIARY_FILE
    if path.exists() or path.is_symlink():
        if path.is_symlink() or not path.is_file() or path.stat().st_size > MAX_FILE_BYTES:
            raise ValueError("dream diary is invalid")
        return path, True
    return path, False


def _insert(existing: str, entry: str, fingerprint: str) -> tuple[str, bool, str | None]:
    body = existing.replace("\r\n", "\n").replace("\r", "\n")
    if fingerprint in body:
        return body, False, "already_present"
    start, end = body.find(START_MARKER), body.rfind(END_MARKER)
    if body.count(START_MARKER) or body.count(END_MARKER):
        if body.count(START_MARKER) != 1 or body.count(END_MARKER) != 1 or start < 0 or end <= start:
            return body, False, "invalid_managed_block"
        return f"{body[:end]}{entry}{body[end:]}", True, None
    section = f"# Dream Diary\n\n{START_MARKER}{entry}\n{END_MARKER}\n"
    return (section if not body.strip() else f"{section}\n{body.lstrip()}"), True, None


def append_dream_diary_entry(
    *,
    workspace_dir: str | Path,
    agent_id: str,
    narrative: str,
    scope: Any = None,
    mode: str = "rem",
    timezone_name: str | None = None,
    now: datetime | None = None,
    workspace_root: str | Path | None = None,
) -> dict[str, Any]:
    """Append one private-agent dream in the host managed block, idempotently."""
    try:
        agent_id = safe_agent_id(agent_id)
        if binding_from_scope(agent_id, scope).scope_type != "agent-private":
            return {"written": False, "code": "not_private_scope"}
        text = _normalize_narrative(narrative)
        if not text:
            return {"written": False, "code": "empty_narrative"}
        workspace = _workspace_path(workspace_dir, workspace_root)
        fingerprint = narrative_fingerprint(text)
        label = "light dream" if mode == "light" else "REM dream"
        entry = f"\n---\n\n*{_date_text(now, timezone_name)}*\n\n{text}\n\n<sub>PLUR1BUS · {label} · {fingerprint}</sub>\n"
        with _DIARY_LOCK:
            path, exists = _diary_path(workspace)
            existing = path.read_text(encoding="utf-8") if exists else ""
            updated, changed, reason = _insert(existing, entry, fingerprint)
            if not changed:
                return {"written": False, "code": reason, "file": DIARY_FILE}
            if len(updated.encode("utf-8")) > MAX_FILE_BYTES:
                return {"written": False, "code": "file_too_large", "file": DIARY_FILE}
            with tempfile.NamedTemporaryFile(
                "w", encoding="utf-8", dir=workspace, prefix=".DREAMS.", delete=False
            ) as handle:
                handle.write(updated)
                temp_name = handle.name
            try:
                os.replace(temp_name, path)
            finally:
                if os.path.exists(temp_name):
                    os.unlink(temp_name)
        return {"written": True, "code": "written", "file": DIARY_FILE}
    except Exception:
        return {"written": False, "code": "write_failed"}
